src/components/menu/__internal__/keyboard-navigation/index.js

```diff
@@ -1,38 +1,26 @@
 import Events from "../../../../__internal__/utils/helpers/events";
-import MenuItem from "../../menu-item";
-
-function characterNavigation(inputString, focusableItems, currentFocusedIndex) {
-  if (!inputString) return currentFocusedIndex;
-
-  const getNodeText = (node) => {
-    if (node instanceof Array) return node.map(getNodeText).join("");
-    if (typeof node === "object" && node)
-      return getNodeText(node.props.children);
-    return node;
-  };
-  const getMenuText = (element) => {
-    if (element.submenu) {
-      return element.submenu;
-    }
-
-    return getNodeText(element.children);
-  };
-
-  const itemsList = focusableItems.map(
-    (item) =>
-      item &&
-      item.type === MenuItem &&
-      getMenuText(item.props) &&
-      getMenuText(item.props).toLowerCase()
-  );
-
-  const matchingItem = itemsList.find(
-    (item) => item && item.startsWith(inputString)
-  );
-
-  const matchingIndex = itemsList.indexOf(matchingItem);
-
-  return matchingIndex === -1 ? currentFocusedIndex : matchingIndex;
+import { MENU_ITEM, SCROLLABLE_BLOCK_PARENT } from "../locators";
+
+function characterNavigation(inputString, focusableItems) {
+  if (!inputString) return -1;
+
+  const getInnerText = (element) =>
+    element?.textContent
+      ?.split("\n")
+      ?.map((text) => text.trim())
+      .join(" ");
+
+  const getMenuText = (element) => getInnerText(element)?.toLowerCase();
+
+  const matchingIndex = focusableItems.findIndex((item) => {
+    return (
+      [MENU_ITEM, SCROLLABLE_BLOCK_PARENT].includes(
+        item?.getAttribute("data-component")
+      ) && getMenuText(item)?.startsWith(inputString?.toLowerCase())
+    );
+  });
+
+  return matchingIndex;
```

Comment:
Very good point, I originally wanted to avoid this as we spread the rest props onto the MenuItems so in theory people could be setting their own `id` on a given item. However, as that's not documented I'm thinking it's not an issue to go with the approach you suggest as it's not a breaking change